File: hub/pet_codex_motion.py
#!/usr/bin/env python3
"""
pet_codex_motion.py — pure signal → Codex-aligned pet motion vocabulary.

Maps live agent/hub evidence onto the standard Codex v2 animation rows:

  idle | running | waiting | failed | review | waving | jumping

(running-right / running-left are locomotion polish; default busy work uses
``running`` row 7.)

This layer is intentionally separate from:
  * pet_manager.derive_mood  — coarse disk-honesty labels for ~/.shannon/pets
  * CompanionMood / PetAnimationState — procedural Canvas mood enums

Honesty rule (same as CompanionMood): observation-only presence may never
claim work. Busy motions require live-capable presence.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Optional

# Codex-aligned motion labels (standard atlas rows 0–8 subset + polish).
MOTION_IDLE = "idle"
MOTION_RUNNING = "running"  # busy work (row 7), not foot-running
MOTION_WAITING = "waiting"
MOTION_FAILED = "failed"
MOTION_REVIEW = "review"
MOTION_WAVING = "waving"  # approval / greeting one-shot
MOTION_JUMPING = "jumping"  # optional celebration polish

# Required vocabulary for Shannon companion responsiveness.
CORE_MOTIONS = frozenset({
    MOTION_IDLE,
    MOTION_RUNNING,
    MOTION_WAITING,
    MOTION_FAILED,
    MOTION_REVIEW,
})

ALL_MOTIONS = frozenset({
    MOTION_IDLE,
    MOTION_RUNNING,
    MOTION_WAITING,
    MOTION_FAILED,
    MOTION_REVIEW,
    MOTION_WAVING,
    MOTION_JUMPING,
    "running-right",
    "running-left",
})

#: Presence values that may drive busy/work motions.
LIVE_PRESENCE = frozenset({"live", "connected", "online"})

#: Status values that mean mid-work (busy), when presence is live.
BUSY_STATUS = frozenset({"active", "mid_task", "mid-task", "running", "working"})

#: Status / flags meaning the agent needs the user (approval, blocked).
WAITING_STATUS = frozenset({
    "blocked", "waiting", "awaiting", "needs_user", "needs-user", "ask",
})

#: Explicit failure outcomes / statuses.
FAILED_OUTCOMES = frozenset({
    "failed", "fail", "error", "errored", "failure", "crash", "crashed",
})

#: Outcomes that map to the review row (inspect completed work).
REVIEW_OUTCOMES = frozenset({
    "review", "done", "success", "succeeded", "completed", "complete", "passed",
})

SLEEPY_AFTER: float = 300.0


@dataclass(frozen=True)
class PetMotionSignals:
    """Evidence tuple for deterministic motion selection.

    Fields mirror gate/card signals without depending on Swift types.
    """

    presence: str = "observed"
    status: str = "idle"
    #: Human ask pending (gate approval queue).
    has_pending_ask: bool = False
    #: Last turn/task outcome: success|failed|review|None.
    last_outcome: Optional[str] = None
    #: Approval celebration window (one-shot, maps to waving).
    just_approved: bool = False
    #: Live entropy collapse (Shannon δ) — surfaces as failed posture.
    entropy_collapse: bool = False
    seconds_since_seen: float = 0.0
    #: Prefer jumping over waving for celebrations when True.
    celebrate_as_jump: bool = False


def _norm(value: Optional[str]) -> str:
    return (value or "").strip().lower().replace(" ", "_")


def is_live_presence(presence: str) -> bool:
    return _norm(presence) in LIVE_PRESENCE


def map_pet_motion(signals: PetMotionSignals) -> str:
    """Map agent/hub signals to a Codex motion label.

    Deterministic pure function. Precedence (top wins):

      1. entropy_collapse + live → failed
      2. just_approved → waving (or jumping)
      3. live + (pending ask | waiting status) → waiting
      4. live + failed outcome → failed
      5. live + busy status → running
      6. live + review/success outcome (not busy) → review
      7. offline / very stale → idle (no sleepy row in Codex atlas)
      8. else idle
    """
    presence = _norm(signals.presence)
    status = _norm(signals.status)
    outcome = _norm(signals.last_outcome) if signals.last_outcome else ""
    live = is_live_presence(presence)

    # 1. Collapse outranks celebration — same honesty as CompanionMood.wary.
    if live and signals.entropy_collapse:
        return MOTION_FAILED

    # 2. Approval one-shot.
    if signals.just_approved:
        return MOTION_JUMPING if signals.celebrate_as_jump else MOTION_WAVING

    # 3–6 only from live-capable presence (observation must not claim work).
    if live:
        if signals.has_pending_ask or status in WAITING_STATUS:
            return MOTION_WAITING
        if outcome in FAILED_OUTCOMES or status in FAILED_OUTCOMES:
            return MOTION_FAILED
        if status in BUSY_STATUS:
            return MOTION_RUNNING
        if outcome in REVIEW_OUTCOMES:
            return MOTION_REVIEW

    # Stale / offline / observed — Codex has no sleepy row; idle is honest.
    return MOTION_IDLE


def motion_claims_work(motion: str) -> bool:
    """True only for motions that assert the agent is doing work."""
    return _norm(motion) in {MOTION_RUNNING, "running-right", "running-left"}


def companion_mood_to_motion(
    mood: str,
    *,
    status: str = "idle",
    has_pending_ask: bool = False,
    last_outcome: Optional[str] = None,
) -> str:
    """Bridge from Shannon procedural mood labels to Codex motion rows.

    Used when a caller already has CompanionMood / pet_manager mood and wants
    an atlas row without re-deriving from raw signals.
    """
    m = _norm(mood)
    st = _norm(status)
    if m in {"wary", "uneasy"}:
        return MOTION_FAILED
    if m in {"happy", "celebrating"}:
        return MOTION_WAVING
    if m in {"alert", "focused", "grinding"}:
        if has_pending_ask or st in WAITING_STATUS:
            return MOTION_WAITING
        if last_outcome and _norm(last_outcome) in FAILED_OUTCOMES:
            return MOTION_FAILED
        return MOTION_RUNNING
    if m in {"review"}:
        return MOTION_REVIEW
    # idle / resting / sleeping / watching / unknown
    if last_outcome and _norm(last_outcome) in REVIEW_OUTCOMES:
        return MOTION_REVIEW
    return MOTION_IDLE

File: Pill/Sources/PillCore/PetCodexMotion.swift
// PetCodexMotion.swift — pure signal → Codex-aligned motion vocabulary.
//
// Maps live agent/hub evidence onto standard Codex v2 atlas rows:
//   idle | running | waiting | failed | review | waving | jumping
//
// Parallel to hub/pet_codex_motion.py (same precedence and honesty rules).
// CompanionMood remains the procedural Canvas mood; this enum drives atlas
// frame selection when a Codex package is available.

import Foundation

/// Codex-compatible animation state for companion surfaces.
public enum PetCodexMotion: String, CaseIterable, Hashable, Sendable {
    case idle
    case running          // busy work (atlas row 7)
    case waiting
    case failed
    case review
    case waving           // approval / greeting one-shot
    case jumping          // optional celebration polish
    case runningRight = "running-right"
    case runningLeft = "running-left"

    /// Motions required by Shannon companion responsiveness.
    public static let core: [PetCodexMotion] = [
        .idle, .running, .waiting, .failed, .review,
    ]

    /// True only for motions that assert the agent is doing work.
    public var claimsWork: Bool {
        switch self {
        case .running, .runningRight, .runningLeft: return true
        default: return false
        }
    }

    /// Foot-walk atlas rows (`running-left` / `running-right`) vs busy-in-place `running`.
    public var isLocomotion: Bool {
        switch self {
        case .runningLeft, .runningRight: return true
        default: return false
        }
    }

    /// Busy-in-place work row (atlas row 7) — not foot locomotion.
    public var isBusyInPlace: Bool { self == .running }

    /// Seconds per foot-walk direction half-cycle on the desktop companion.
    public static let locomotionStrideSeconds: Double = 2.4

    /// Atlas playback motion for a resolved base motion.
    ///
    /// - Busy-in-place `running` can intentionally become `running-left` /
    ///   `running-right` for desktop walk polish (phase-stable on `tSeconds`).
    /// - Already-directed locomotion is preserved.
    /// - Non-work motions are unchanged (never invent walk for idle/wait/fail).
    /// - `allowLocomotion: false` keeps busy-in-place (procedural / Reduce Motion).
    public static func atlasPlaybackMotion(
        base: PetCodexMotion,
        tSeconds: Double,
        agentSeed: String = "",
        allowLocomotion: Bool = true
    ) -> PetCodexMotion {
        switch base {
        case .runningLeft, .runningRight:
            return base
        case .running:
            guard allowLocomotion else { return .running }
            return locomotionDirection(tSeconds: tSeconds, agentSeed: agentSeed)
        default:
            return base
        }
    }

    /// Phase-stable left/right foot-walk from wall time + agent seed.
    public static func locomotionDirection(
        tSeconds: Double,
        agentSeed: String = ""
    ) -> PetCodexMotion {
        let period = max(0.5, locomotionStrideSeconds)
        let t = max(0.0, tSeconds)
        var hash = 0
        for b in agentSeed.utf8 {
            hash = hash &* 31 &+ Int(b)
        }
        // Bucket flips every `period` seconds; seed biases start side.
        let bucket = Int(floor(t / period)) &+ (hash & 1)
        return (bucket & 1) == 0 ? .runningRight : .runningLeft
    }

    /// Evidence inputs for deterministic motion selection.
    public struct Signals: Sendable, Equatable {
        public var presence: AgentPresence
        public var status: AgentRunStatus
        public var hasPendingAsk: Bool
        public var lastOutcome: String?
        public var justApproved: Bool
        public var entropyCollapse: Bool
        public var celebrateAsJump: Bool

        public init(
            presence: AgentPresence = .observed,
            status: AgentRunStatus = .idle,
            hasPendingAsk: Bool = false,
            lastOutcome: String? = nil,
            justApproved: Bool = false,
            entropyCollapse: Bool = false,
            celebrateAsJump: Bool = false
        ) {
            self.presence = presence
            self.status = status
            self.hasPendingAsk = hasPendingAsk
            self.lastOutcome = lastOutcome
            self.justApproved = justApproved
            self.entropyCollapse = entropyCollapse
            self.celebrateAsJump = celebrateAsJump
        }
    }

    private static let failedOutcomes: Set<String> = [
        "failed", "fail", "error", "errored", "failure", "crash", "crashed",
    ]
    private static let reviewOutcomes: Set<String> = [
        "review", "done", "success", "succeeded", "completed", "complete", "passed",
    ]

    /// Pure mapping. Precedence matches hub/pet_codex_motion.py.
    public static func map(_ signals: Signals) -> PetCodexMotion {
        let live = signals.presence.canBeBusy
        let outcome = (signals.lastOutcome ?? "")
            .trimmingCharacters(in: .whitespacesAndNewlines)
            .lowercased()

        // 1. Collapse outranks celebration.
        if live, signals.entropyCollapse { return .failed }

        // 2. Approval one-shot.
        if signals.justApproved {
            return signals.celebrateAsJump ? .jumping : .waving
        }

        // 3–6 only from live-capable presence.
        if live {
            if signals.hasPendingAsk || signals.status == .blocked {
                return .waiting
            }
            if failedOutcomes.contains(outcome) { return .failed }
            if signals.status == .active || signals.status == .midTask {
                return .running
            }
            if reviewOutcomes.contains(outcome) { return .review }
        }

        return .idle
    }

    /// Bridge from procedural CompanionMood (+ optional status cues).
    public static func from(
        mood: CompanionMood,
        status: AgentRunStatus = .idle,
        hasPendingAsk: Bool = false,
        lastOutcome: String? = nil
    ) -> PetCodexMotion {
        switch mood {
        case .wary:
            return .failed
        case .happy:
            return .waving
        case .alert:
            if hasPendingAsk || status == .blocked { return .waiting }
            if let o = lastOutcome?.lowercased(), failedOutcomes.contains(o) {
                return .failed
            }
            return .running
        case .idle, .sleepy:
            if let o = lastOutcome?.lowercased(), reviewOutcomes.contains(o) {
                return .review
            }
            return .idle
        }
    }

    /// Resolve from a live activity snapshot.
    public static func resolve(
        for agent: AgentActivitySnapshot,
        now: Date = Date(),
        approvedAt: Date? = nil,
        hasPendingAsk: Bool = false,
        lastOutcome: String? = nil,
        entropyDelta: Double? = nil,
        collapseThreshold: Double = CompanionMood.defaultCollapseThreshold
    ) -> PetCodexMotion {
        let collapse: Bool = {
            guard agent.presence == .live, let d = entropyDelta else { return false }
            return d <= collapseThreshold
        }()
        let approved: Bool = {
            guard let at = approvedAt else { return false }
            let age = now.timeIntervalSince(at)
            return age >= 0 && age <= CompanionMood.happyDuration
        }()
        return map(Signals(
            presence: agent.presence,
            status: agent.status,
            hasPendingAsk: hasPendingAsk,
            lastOutcome: lastOutcome,
            justApproved: approved,
            entropyCollapse: collapse
        ))
    }
}
